lib/event_store/aggregate.ex

```diff
@@ -43,59 +56,95 @@ defmodule EventStore.Aggregate do
     quote location: :keep do
       use GenServer
 
+      import unquote(__MODULE__)
+
       @behaviour unquote(__MODULE__)
       @before_compile unquote(__MODULE__)
 
-      def init(%HLClock.Timestamp{} = id) do
-        {:ok, %{id: id, sequence: 0, state: nil}}
-      end
+      def initial_state, do: %{}
 
-      def handle_call(:get_state, _from, state), do: {:reply, state, state}
+      def eval_command(_, _), do: []
 
-      def handle_call(:snapshot, _from, agg), do: {:ok, to_snapshot(agg), agg}
+      def apply_event(prev, _), do: prev
 
-      def handle_call(:update_state, _from, agg),
-        do: {:reply, :ok, update_aggregate(agg)}
+      def prepare_snapshot(s), do: s
 
-      def handle_call({:eval_command, command}, _from, agg) do
-        # command implies knowledge of events we haven't seen yet, update before
-        # processing
-        agg = case command.sequence > agg.sequence do
-                true -> update_aggregate(agg)
-                false -> agg
-              end
+      def use_snapshot(_, snapshot), do: snapshot.body
 
-        {:reply, eval_command(agg.state, command), agg}
+      defoverridable [
+        initial_state: 0,
+        eval_command: 2,
+        apply_event: 2,
+        prepare_snapshot: 1,
+        use_snapshot: 2
+      ]
+
+      def init(%HLClock.Timestamp{} = id) do
+        send(self(), :initialize)
+        {:ok, id |> create_aggregate()}
       end
 
-      def handle_call({:apply_events, []}, _from, state),
-        do: {:reply, :ok, state}
-      def handle_call({:apply_events, events}, _from, state) do
-        {:reply, :ok, apply_events(state, events)}
+      def handle_call(:get_state, _from, agg), do: {:reply, agg.state, agg}
+      def handle_call({:get_state, seq}, _from, agg) do
+        {:reply, at_sequence(agg, seq), agg}
+      end
+      def handle_call(:get_snapshot, _from, agg) do
+        {:reply, to_snapshot(agg), agg}
+      end
+      def handle_call({:eval_command, command}, _from, agg) do
+        {:reply, :ok, handle_command(agg, command)}
       end
 
-      def eval_command(_, _), do: []
+      def handle_info(:initialize, agg), do: {:noreply, update_aggregate(agg)}
 
-      def apply_event(prev, _), do: prev
+      defp handle_command(agg, com) do
+        with agg                   <- command_update(agg, com),
+             evs when is_list(evs) <- eval_command(agg, com),
+               %Aggregate{} = agg  <- handle_events(agg, com, evs) do
+          agg
+        end
+      end
 
-      def use_snapshot(_, snapshot), do: snapshot.body
+      defp command_update(%{sequence: a} = agg, %{sequence: c}) when c > a,
+        do: update_aggregate(agg)
+      defp command_update(agg, _), do: agg
+
+      def handle_events(agg, command, events) do
+        case Store.commit_events!(events) do
```

Comment:
If its a `!` function then it shouldn't be returning a match it should just raise an exception.